reporter/aspec/src/main/java/org/itsallcode/openfasttrace/report/aspec/ASpecReport.java

```diff
@@ -0,0 +1,362 @@
+     * @param context configuration options
+     */
+    public ASpecReport( final Trace trace, final ReporterContext context )
+    {
+        this.trace = trace;
+        this.newline = context.getSettings().getNewline();
+        this.xmlOutputFactory = XMLOutputFactory.newFactory();
+    }
+
+
+    @Override
+    public void renderToStream( final OutputStream outputStream )
+    {
+        final XMLStreamWriter xmlWriter = createXmlWriter( new OutputStreamWriter( outputStream ) );
+        final IndentingXMLStreamWriter indentingXmlWriter = new IndentingXMLStreamWriter( xmlWriter );
+        final Map<String, List<LinkedSpecificationItem>> items = groupByDoctype( this.trace.getItems().stream() );
+
+        LOG.info( "aspec starting" );
+        try( indentingXmlWriter )
+        {
+            writeOutput( indentingXmlWriter, items );
+        } catch( final XMLStreamException e )
+        {
+            throw new ExporterException( "Generating document", e );
+        }
+    }
+
+    private XMLStreamWriter createXmlWriter( final Writer writer )
+    {
+        try
+        {
+            return this.xmlOutputFactory.createXMLStreamWriter( writer );
+        } catch( final XMLStreamException e )
+        {
+            throw new ExporterException( "Error creating xml stream writer for writer " + writer, e );
+        }
+    }
+
+    private Map<String, List<LinkedSpecificationItem>> groupByDoctype(
+            final Stream<LinkedSpecificationItem> itemStream )
+    {
+        return itemStream.collect(
+                groupingBy( LinkedSpecificationItem::getArtifactType, LinkedHashMap::new, toList() ) );
+    }
+
+    private void writeOutput( final XMLStreamWriter writer, Map<String, List<LinkedSpecificationItem>> items ) throws XMLStreamException
+    {
+        writer.writeStartDocument( "UTF-8", "1.0" );
+        writer.writeStartElement( "specdocument" );
+
+        for( final Map.Entry<String, List<LinkedSpecificationItem>> entry : items.entrySet() )
+        {
+            final String doctype = entry.getKey();
+            final List<LinkedSpecificationItem> specItems = entry.getValue();
+            writeItems( writer, doctype, specItems );
+        }
+
+        writer.writeEndElement();
+        writer.writeEndDocument();
+    }
+
+    private void writeItems( final XMLStreamWriter writer, final String doctype, final List<LinkedSpecificationItem> specItems )
+            throws XMLStreamException
+    {
+        LOG.finest( () -> "Writing " + specItems.size() + " items with doctype " + doctype );
+        writer.writeStartElement( "specobjects" );
+        writer.writeAttribute( "doctype", doctype );
+        for( final LinkedSpecificationItem item : specItems )
+        {
+            writeItem( writer, item );
+        }
+        writer.writeEndElement();
+    }
+
+    private void writeItem( final XMLStreamWriter writer, final LinkedSpecificationItem item ) throws XMLStreamException
```

Comment:
Please split this method into section by extracting cohesive parts to methods with speaking names.